### ISSUE TITLE
New exercise: Leap

### ISSUE THREAD

**Issue by kytrinyx** [exercism/chapel#11]
I'd like to generate this exercise, which is a very simple exercise about boolean logic.

Doing a simple search/replace into the HelloWorld test, I get the following:

```chapel
proc test_something(test : borrowed Test) throws {
  test.assertEqual(is_leap_year(2015), false);
}
```

Do you use `assertEqual` to test true and false, or is there a different test method?

What would the idiomatic name be for the function under test? (here: `is_leap_year`)

Are there any other changes you'd want to make?

**Comment by lucaferranti** [comment]
Great, I also started yesterday writing the leap year and I had indeed a few questions.

Answering your questions first
1. [Chapel style guidelines](https://chapel-lang.org/docs/developer/bestPractices/StandardModuleStyle.html) recommends to use `camelCase` for functions, so it should probably be `isLeapYear` (btw, chapel has such a function already in the `Time` module, that is not loaded by default, so it should not cause problems). 

I think one reason why my brain didn't think to point this out in the `two-fer` exercise is that imho  `test_another_name_given` is more readable than `testAnotherNameGiven`.

I would say for functions under test, `e.g. isLeapYear` we could follow the guidelines and use camelCase, for the test-case functions, I would almost be tempted to recommend `snake_case`, it just feals more readable if the name is long, thoughts on that?

2. Chapel has `test.assertTrue` and `test.assertFalse`. They could be nice to use, but if makes the automatic generation significantly harder, `test.assertEqual` would be fine too.
 
Regarding my questions:

- Is it so that the concepts, prerequisites of each exercise are not fixed in the specs, but the author has freedom of what to add? I browsed some other repositories and I see some variance.

- Are there some "official test cases" associated with each exercise in the specifications and metadata or does each author come up with their own test-cases?

**Comment by kytrinyx** [comment]
We should definitely follow Chapel style guidelines. For the tests I do feel like `snake_case` is easier to read for long names, but I think it's better to give people good habits in this case and use whatever Chapel suggests is better.

`assertTrue` and `assertFalse` is great and not at all difficult to do in terms of templates.

> Is it so that the concepts, prerequisites of each exercise are not fixed in the specs, but the author has freedom of what to add? I browsed some other repositories and I see some variance.

TL;DR: yes.

The more fleshed out answer is that there are two types of exercise on Exercism: concept exercises and practice exercises. When spinning up a new track, we make a bucket of practice exercises, and then when the track is launched and getting some traction, we advise creating concept exercises (they're harder to get right and require some careful thinking because they involve creating a structured syllabus).

For practice exercises we are really just providing a playground for people to start experimenting with the language. Often practice exercises have several possible approaches, which lets people play around and learn based on experimentation and reading other people's solutions as well.

With concept exercises the opposite is true: we typically have a specific thing we're trying to teach, and want people to take a specific approach to solving them.

So in other words with these exercises that we're creating it's up to the language and author and what makes sense or seems fun.

> Are there some "official test cases" associated with each exercise in the specifications and metadata or does each author come up with their own test-cases?

Sort of! We have a whole repository full of specifications for practice exercises (https://github.com/exercism/problem-specifications), but none of the test cases are officially mandated, they're more like a deli where you can go pick whatever makes sense in the language. It's what I'm basing the template generator on.

The `bin/configlet sync --update --tests --exercise leap` will show you the specification data for each test and ask you to make a decision about which tests to include and which to exclude.

I've just been including all of them for now, because most of the initial exercises are pretty straight forward, but we'll get to some exercises where we'll want to make a decision about whether to include certain test cases.

**Comment by lucaferranti** [comment]
not sure if I'm doing something wrong, but running

```
bin/configlet sync --update --tests --exercise leap
```

indeed did ask me what test cases to include, e.g.

```
{
  "uuid": "c30331f6-f9f6-4881-ad38-8ca8c12520c1",
  "description": "year divisible by 200, not divisible by 400 in common year",
  "property": "leapYear",
  "input": {
    "year": 1800
  },
  "expected": false
}
```

but the generated `meta/tests.toml` file only has


```
[c30331f6-f9f6-4881-ad38-8ca8c12520c1]
description = "year divisible by 200, not divisible by 400 in common year"
```

are the (input, output) pairs stored somewhere else?

**Comment by kytrinyx** [comment]
No, you're not doing anything wrong. If you say "y" does it continue and ask you about the next test and the next test?

(I opened a PR here with what I got on my end: https://github.com/exercism/chapel/pull/17)

**Comment by lucaferranti** [comment]
> No, you're not doing anything wrong. If you say "y" does it continue and ask you about the next test and the next test?

yes, exactly and I pressed `y` to all but the toml file only had the description

**Comment by lucaferranti** [comment]
with a quick look, the generated chapel code looks good, but where does your generator read what the input and expected output should be?

**Comment by kytrinyx** [comment]
Oh! I see what you mean. Yes. It gets the data from https://github.com/exercism/problem-specifications/blob/main/exercises/leap/canonical-data.json

**Comment by lucaferranti** [comment]
I see cool, but those json data are not written by default into the `tests.toml` (or even downloaded)? Is it possible to tell configlet to download those? I guess this would be useful for the generator, or does it get the data with an HTTP request?

**Comment by kytrinyx** [comment]
Configlet stashes a copy of the `problem-specifications` repository in the user's library caches, and you can pass `--offline` to configlet to have it not do network request.

The `tests.toml` is where it would say `include = false` if you decided that you wanted to skip a test for some reason (e.g. if the test uses really big numbers and it would make the exercise too hard to rely on some "BigNumeric" implementation, you might want to skip that test).

So for the sake of generating I use the `tests.toml` to decide which tests to include, and then the `canonical-data.json` from the problem-specifications to get the actual data.
